### PR TITLE
feat: User Authentication using JWT and Mongo DB

### PR DESCRIPTION
**Pull request by prudentbird-dev**
# Description

This PR implements user authentication features using JWT and MongoDB.

# Changes proposed

- **Sign Up**: Added `POST ROUTE` `api/auth/register` endpoint to register new users.
  - Input: Email, Password, First Name, Last Name.
  - Features password hashing with bcrypt and stores user details in MongoDB.
  - Outputs a JWT token and user profile upon successful sign-up.

- **JWT Service**: Added `JWT Service`.
  - Generates, saves and validates JWT tokens
  - Implement refresh and access tokens to persist user session for up to a month

- **Email Service**: Added `Email Service`. 
  - Sends welcome and verification mail on registration.

- **User Service**: Added `User Service`.
  - Provides methods to interact with user data and perform CRUD operations

- **Verify Email**: Added `POST ROUTE` `api/auth/verify-email` endpoint to verify email address.
  - Input: Email verificiation token `token: "<sample token>"`.
  - Validates token and verifies the user.

- **Error Handler**: Added `Error Handling Middleware`.
  - Sends a custom error response during unexpected outcome or event failure

- Added exportDirStructure function
- Installed necessary dependencies
- Implemented pagination and toJSON plugins for models
- Implemented Role Based Access Control
- Added authGuard, adminGuard and superAdminGuard

# Usage

- Register a user on `POST api/auth/register`.<br>
  - Sample request using POSTMAN<br><br>
  ```
    {
      "email": "{{$randomEmail}}",
      "password": "{{$randomPassword}}",
      "firstName": "{{$randomFirstName}}",
      "lastName": "{{$randomLastName}}"
    }
  ``` 

- Verify a user on `POST api/auth/verify-email`.<br>
  - Sample request using POSTMAN<br><br>
  ```
    {
      "token": "sample token as string",
    }
  ``` 

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos

image

### CODE REVIEW

File: src/controllers/auth.controller.ts
Comment (shalomtaiwo) [inline src/controllers/auth.controller.ts:1]: It should only be an email and password for signup.

And although the verification is good, it should also not be compulsory. Users should still be logged in and be able to use the workspace. However, they would not be able to publish any page unless they verify their email.

File: src/services/email.service.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * Send a welcome email to a new user.
+   * @param to - Recipient email address.
+   * @param firstName - The first name of the new user.
+   * @returns A promise that resolves when the email is sent.
+   */
+  sendWelcomeEmail: async (to: string, firstName: string): Promise<void> => {
+    const subject = "Welcome to Reavpages";
+    const text = `Hello ${firstName},
```
Comment (shalomtaiwo) [inline src/services/email.service.ts:63]: Instead of ${firstName} use ${email} - But only for this welcome email.

File: src/types/user.types.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+   * The unique identifier of the user.
+   * @example "60c72b2f9b1e8b001f8e4e67"
+   */
+  id: string;
+
+  /**
+   * The first name of the user.
+   * @example "John"
+   */
+  firstName: string;
```
Comment (shalomtaiwo) [inline src/types/user.types.ts:17]: Should be null by default

File: src/types/user.types.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * The last name of the user.
+   * @example "Doe"
+   */
+  lastName: string;
```
Comment (shalomtaiwo) [inline src/types/user.types.ts:23]: should be null by default

**Review by shalomtaiwo — DISMISSED** [review]

File: src/controllers/auth.controller.ts
Comment (prudentbird-dev) [reply src/controllers/auth.controller.ts:1]: A user can login without verifying email, verification isn't compulsory and there are no restrictions yet

Effecting the requested change in the register Schema removes the firstName field and that affects the welcome message, sent via mail. 

Would the firstName field be provided later on via an onboarding process after registration that updates the user with the required fields?

File: src/services/auth.service.ts
[the remark this answers is not in the document]
Comment (prudentbird-dev) [reply src/services/auth.service.ts:1]: Duly noted, this resolves an answer to my initial question regarding the changes in the registration Schema

File: src/services/email.service.ts (hunk 1 above)
Comment (prudentbird-dev) [reply src/services/email.service.ts:63]: Duly noted, this resolves an answer to my initial question regarding the changes in the registration Schema.

File: src/types/user.types.ts (hunk 2 above)
Comment (prudentbird-dev) [reply src/types/user.types.ts:17]: The user firstName is undefined/null by default. I presume you're referring to the user model, I would effect the changes requested.

File: src/types/user.types.ts (hunk 3 above)
Comment (prudentbird-dev) [reply src/types/user.types.ts:23]: The user lastName is undefined/null by default. I presume you're referring to the user model, I would effect the changes requested.

File: src/controllers/auth.controller.ts
Comment (shalomtaiwo) [reply src/controllers/auth.controller.ts:1]: The user would be able to add their full details in the account settings.

Comment (prudentbird-dev) [reply src/controllers/auth.controller.ts:1]: Okay, changes requested implemented in the last commit!

**Review by shalomtaiwo — APPROVED** [review]
Well done

**Pull request merged**